knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go

```diff
@@ -1,51 +1,187 @@
 	mf "github.com/manifestival/manifestival"
+	consolev1 "github.com/openshift/api/console/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	consoleroute "github.com/openshift/console-operator/pkg/console/subresource/route"
+	consoleutil "github.com/openshift/console-operator/pkg/console/subresource/util"
+	k8sv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	servingv1alpha1 "knative.dev/serving-operator/pkg/apis/serving/v1alpha1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-const defaultKnConsoleCLIDownload = "deploy/resources/console_cli_download_kn.yaml"
+const (
+	knDownloadServer                    = "kn-download-server"
+	knConsoleCLIDownloadDeployRoute     = "kn-cli-downloads"
+	knConsoleCLIDownloadDeployNamespace = "kn-cli-downloads"
+)
 
 var log = common.Log.WithName("consoleclidownload")
 
-// Create creates ConsoleCLIDownload for kn CLI download links
-func Create(instance *servingv1alpha1.KnativeServing, apiclient client.Client) error {
-	log.Info("Creating ConsoleCLIDownload CR for kn")
-	manifest, err := mfc.NewManifest(manifestPath(), apiclient, mf.UseLogger(log.WithName("mf")))
+// Create deploy deployment and CR for kn ConsoleCLIDownload
+func Create(instance *servingv1alpha1.KnativeServing, apiclient client.Client, scheme *runtime.Scheme) error {
+	addToScheme(scheme)
+	if err := createKnDeployment(apiclient, scheme); err != nil {
+		return err
+	}
+
+	if err := createCR(apiclient); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// createKnDeployment creates required resources viz Namespace, Deployment, Service, Route
+// which will serve kn cross platform binaries within cluster
+func createKnDeployment(apiclient client.Client, scheme *runtime.Scheme) error {
+	log.Info("Creating kn ConsoleCLIDownload deployment")
+	manifest, err := mfc.NewManifest(rawManifestKnDownloadResources(), apiclient)
 	if err != nil {
-		return fmt.Errorf("failed to read ConsoleCLIDownload manifest: %w", err)
+		return fmt.Errorf("failed to read kn ConsoleCLIDownload deployment manifest: %w", err)
 	}
+
+	manifest, err = manifest.Transform(replaceKnCLIArtifactsImage(os.Getenv("IMAGE_KN_CLI_ARTIFACTS"), scheme))
+	if err != nil {
+		return fmt.Errorf("failed to transform kn ConsoleCLIDownload deployment image: %w", err)
+	}
+
 	if err := manifest.Apply(); err != nil {
-		return fmt.Errorf("failed to apply ConsoleCLIDownload manifest: %w", err)
+		return fmt.Errorf("failed to apply kn ConsoleCLIDownload download resources manifest: %w", err)
+	}
+
+	return nil
+}
+
+// createCR creates kn ConsoleCLIDownload CR. It finds the route serving kn binaries and
+// populates ConsoleCLIDownload object accordingly and creates it.
+func createCR(apiclient client.Client) error {
+	route := &routev1.Route{}
+	// Waiting for the deployment/route to appear. TODO: ideally this should be a watch
+	time.Sleep(time.Second * 10)
```

Comment:
Added using `k8s.io/apimachinery/pkg/util/wait.Pollimmediate` with interval of 2 secs and timeout of 20 secs.

Reply:
That's not a watch though. We cannot block the reconcile loop for longer than maybe a second or so. This will block the reconcile loop for 20seconds worst case.

Reply:
A watch would need to be setup via setting up a literal `Watch`, listening to the events of the resources and triggering a reconcile once something changes.